pkg/cmd/admin/registry/registry.go

```diff
@@ -243,150 +266,265 @@ func RunCmdRegistry(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg
-					{
-						Name:  "registry",
-						Image: image,
-						Ports: ports,
-						Env:   env.List(),
-						VolumeMounts: []kapi.VolumeMount{
-							{
-								Name:      "registry-storage",
-								MountPath: cfg.Volume,
-							},
-						},
-						SecurityContext: &kapi.SecurityContext{
-							Privileged: &mountHost,
-						},
-						LivenessProbe:  livenessProbe,
-						ReadinessProbe: readinessProbe,
+		servingCert = data
+	}
+	if len(cfg.ServingKeyPath) > 0 {
+		data, err := ioutil.ReadFile(cfg.ServingKeyPath)
+		if err != nil {
+			return fmt.Errorf("registry does not exist; could not load TLS private key file %q: %v", cfg.ServingKeyPath, err)
+		}
+		servingCert = data
+	}
+
+	env := app.Environment{}
+	env.Add(secretEnv)
+
+	healthzPort := defaultPort
+	if len(ports) > 0 {
+		healthzPort = ports[0].ContainerPort
+		env["REGISTRY_HTTP_ADDR"] = fmt.Sprintf(":%d", healthzPort)
+		env["REGISTRY_HTTP_NET"] = "tcp"
+	}
+	secrets, volumes, mounts, extraEnv, tls, err := generateSecretsConfig(cfg, namespace, servingCert, servingKey)
+	if err != nil {
+		return err
+	}
+	env.Add(extraEnv)
+
+	livenessProbe := generateLivenessProbeConfig(healthzPort, tls)
+	readinessProbe := generateReadinessProbeConfig(healthzPort, tls)
+
+	mountHost := len(cfg.HostMount) > 0
+	podTemplate := &kapi.PodTemplateSpec{
+		ObjectMeta: kapi.ObjectMeta{Labels: label},
+		Spec: kapi.PodSpec{
+			NodeSelector: nodeSelector,
+			Containers: []kapi.Container{
+				{
+					Name:  "registry",
+					Image: image,
+					Ports: ports,
+					Env:   env.List(),
+					VolumeMounts: append(mounts, kapi.VolumeMount{
+						Name:      "registry-storage",
+						MountPath: cfg.Volume,
+					}),
+					SecurityContext: &kapi.SecurityContext{
+						Privileged: &mountHost,
 					},
+					LivenessProbe:  livenessProbe,
+					ReadinessProbe: readinessProbe,
 				},
-				Volumes: []kapi.Volume{
+			},
+			Volumes: append(volumes, kapi.Volume{
+				Name:         "registry-storage",
+				VolumeSource: kapi.VolumeSource{},
+			}),
+		},
+	}
+	if mountHost {
+		podTemplate.Spec.Volumes[len(podTemplate.Spec.Volumes)-1].HostPath = &kapi.HostPathVolumeSource{Path: cfg.HostMount}
+	} else {
+		podTemplate.Spec.Volumes[len(podTemplate.Spec.Volumes)-1].EmptyDir = &kapi.EmptyDirVolumeSource{}
+	}
+
+	objects := []runtime.Object{}
+	for _, s := range secrets {
+		objects = append(objects, s)
+	}
+	if needServiceAccount {
+		objects = append(objects,
+			&kapi.ServiceAccount{ObjectMeta: kapi.ObjectMeta{Name: cfg.ServiceAccount}},
+			&authapi.ClusterRoleBinding{
+				ObjectMeta: kapi.ObjectMeta{Name: fmt.Sprintf("registry-%s-role", cfg.Name)},
+				Subjects: []kapi.ObjectReference{
 					{
-						Name:         "registry-storage",
-						VolumeSource: kapi.VolumeSource{},
+						Kind:      "ServiceAccount",
+						Name:      cfg.ServiceAccount,
+						Namespace: namespace,
 					},
 				},
+				RoleRef: kapi.ObjectReference{
+					Kind: "ClusterRole",
+					Name: "system:registry",
```

Comment:
Same argument as with router - router creation shouldn't depend on
constants on server side.  If bootstrap was presented as an API
constants package maybe.  It's also the same reason we don't use a
constant for kind "Pod".